### PR TITLE
Added padding at the begining to avoid tf STFT reconstruction error

### PR DESCRIPTION
**Pull request by romi1502**
# Fix tensorflow STFT reconstruction

## Description

Pad the beginning of the waveform with 0 to avoid tensorflow inverse STFT reconstruction error (as done for librosa backend)

## Documentation link and external references
Should fix issue #437

### CODE REVIEW

File: spleeter/model/__init__.py [hunk 1]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
         spec_name = self.spectrogram_name
 
         if stft_name not in self._features:
+            # pad input with a frame of zeros
+            waveform =  tf.concat([
+                            tf.zeros((self._frame_length, self._n_channels)),
+                            self._features['waveform']
+                            ],
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:282]: 34

Comment (joaopedrpgqo) [reply spleeter/model/__init__.py:282]: 15

File: spleeter/model/__init__.py [hunk 2]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
         spec_name = self.spectrogram_name
 
         if stft_name not in self._features:
+            # pad input with a frame of zeros
+            waveform =  tf.concat([
+                            tf.zeros((self._frame_length, self._n_channels)),
+                            self._features['waveform']
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:281]: 15

File: spleeter/model/__init__.py [hunk 3]
```diff
@@ -341,7 +348,7 @@ def _inverse_stft(self, stft_t, time_crop=None):
         reshaped = tf.transpose(inversed)
         if time_crop is None:
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:349]: 50

File: spleeter/model/__init__.py [hunk 4]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
...
+                            self._features['waveform']
+                            ],
+                            0
+                        )
             stft_feature = tf.transpose(
                 stft(
-                    tf.transpose(self._features['waveform']),
+                    tf.transpose(waveform),
                     self._frame_length,
                     self._frame_step,
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:289]: 25

File: spleeter/model/__init__.py [hunk 5]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
...
+            # pad input with a frame of zeros
+            waveform =  tf.concat([
+                            tf.zeros((self._frame_length, self._n_channels)),
+                            self._features['waveform']
+                            ],
+                            0
+                        )
             stft_feature = tf.transpose(
                 stft(
-                    tf.transpose(self._features['waveform']),
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:280]: 26

File: spleeter/model/__init__.py [hunk 6]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
         spec_name = self.spectrogram_name
 
         if stft_name not in self._features:
+            # pad input with a frame of zeros
+            waveform =  tf.concat([
+                            tf.zeros((self._frame_length, self._n_channels)),
+                            self._features['waveform']
+                            ],
+                            0
+                        )
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:284]: 5

File: spleeter/model/__init__.py [hunk 7]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
         spec_name = self.spectrogram_name
 
         if stft_name not in self._features:
+            # pad input with a frame of zeros
+            waveform =  tf.concat([
+                            tf.zeros((self._frame_length, self._n_channels)),
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:280]: 25

File: spleeter/model/__init__.py [hunk 8]
```diff
@@ -275,9 +275,16 @@ def _build_stft_feature(self):
         spec_name = self.spectrogram_name
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:275]: 100

File: spleeter/model/__init__.py [hunk 9]
```diff
@@ -341,7 +348,7 @@ def _inverse_stft(self, stft_t, time_crop=None):
         reshaped = tf.transpose(inversed)
         if time_crop is None:
             time_crop = tf.shape(self._features['waveform'])[0]
-        return reshaped[:time_crop, :]
+        return reshaped[self._frame_length:self._frame_length+time_crop, :]
```
Comment (joaopedrpgqo) [inline spleeter/model/__init__.py:351]: 40

**Comment by faroit** [comment]
@romi1502 just jumping in: such a change might have a minor effect on the MUSDB18 evaluation scores. Would [your regression tests](https://github.com/deezer/spleeter/blob/master/tests/test_eval.py) catch this? Maybe you recompute these to check if they are still the same...

**Comment by mmoussallam** [comment]
Hi @faroit indeed it will change a bit the scores. The regression test failing is actually here to remind us of that. overall this should be negligible, but we'll run the full eval and quantify exactly how big the impact is.

**Pull request merged**